File: 2020/d5.py
""" https://adventofcode.com/2020/day/5 """

from functools import reduce
from typing import List


def part1(boarding_passes: List[str]) -> int:
    """ O(n) solution """

    return max([get_row(bp) * 8 + get_column(bp) for bp in boarding_passes])


def part2(boarding_passes: List[str]) -> int:
    """ O(n) solution """

    max_seats = 2 ** len(boarding_passes[0])
    seats = [False] * max_seats
    hi_seat, lo_seat = 0, max_seats
    for bp in boarding_passes:
        id = get_row(bp) * 8 + get_column(bp)
        hi_seat = max(hi_seat, id)
        lo_seat = min(lo_seat, id)
        seats[id] = True

    for i in range(lo_seat, hi_seat):
        if not seats[i]:
            return i

    return -1


def get_row(bp: str) -> int:
    return int(reduce(lambda x, y: x + str(y.count("B")), bp[:7], ""), 2)


def get_column(bp: str) -> int:
    return int(reduce(lambda x, y: x + str(y.count("R")), bp[-3:], ""), 2)


if __name__ == "__main__":
    TEST = [line.strip() for line in open("tests/d5.txt", "r")]
    PUZZLE = [line.strip() for line in open("puzzles/d5.txt", "r")]

    assert part1(TEST) == 820
    assert part2(TEST) == 120

    print(f"Part 1: {part1(PUZZLE)}")
    print(f"Part 2: {part2(PUZZLE)}")

File: 2021/src/aoc_2021/d5.clj
(ns aoc-2021.d5 (:require
                 [clojure.string :as str]))

(defn- parse-input [path]
  (->> path
       (slurp)
       (str/split-lines)
       (map (fn [line] (->> line
                            (re-matcher #"(\d+),(\d+) -> (\d+),(\d+)")
                            (re-find)
                            (rest)
                            (map #(Integer/parseInt %))
                            (partition 2))))))

(defn- update-vent-map [vent-map point]
  (update vent-map point (fnil inc 0)))

(defn- calc-straight-line [[x1 y1] [x2 y2]]
  (cond
    (= x1 x2) (map vector
                   (repeat x1)
                   (range (min y1 y2) (inc (max y1 y2))))
    (= y1 y2) (map vector
                   (range (min x1 x2) (inc (max x1 x2)))
                   (repeat y1))))

(defn- calc-diagonal-line [[x1 y1] [x2 y2]]
  (cond
    (or (and (< x1 x2) (> y1 y2))
        (and (> x1 x2) (< y1 y2))) (map vector
                                        (range (min x1 x2) (inc (max x1 x2)))
                                        (range (max y1 y2) (dec (min y1 y2)) -1))
    (or (and (< x1 x2) (< y1 y2))
        (and (> x1 x2) (> y1 y2))) (map vector
                                        (range (min x1 x2) (inc (max x1 x2)))
                                        (range (min y1 y2) (inc (max y1 y2))))))

(defn- calc-mixed-line [point-1 point-2]
  (or (calc-straight-line point-1 point-2)
      (calc-diagonal-line point-1 point-2)))

(defn- draw-line [vent-map points calc-fun]
  (->> points
       (apply calc-fun)
       (reduce update-vent-map vent-map)))

(defn- count-overlaps [lines calc-fun]
  (->> lines
       (reduce #(draw-line %1 %2 calc-fun) {})
       (filter #(> (val %) 1))
       (count)))

(defn part-1
  "At how many points do at least two lines overlap?"
  [lines]
  (count-overlaps lines calc-straight-line))

(defn part-2
  "At how many points do at least two lines overlap?"
  [lines]
  (count-overlaps lines calc-mixed-line))

;; Puzzle: https://adventofcode.com/2021/day/5

(def test-input (parse-input "resources/tests/d5.txt"))
(def input (parse-input "resources/puzzles/d5.txt"))

(= (part-1 test-input) 5)
(time (part-1 input))

(= (part-2 test-input) 12)
(time (part-2 input))
